### PR TITLE
Implement technical specification #1019

### PR DESCRIPTION
**Pull request by maslenitsa93**
#1019

### CODE REVIEW

**Comment by afalaleev** [comment]
I will review this PR after completing of PR #1031

**Comment by maslenitsa93** [comment]
@zxcat, conflicts fixed. Review, please.

File: libraries/chain/include/golos/chain/worker_proposal_objects.hpp [hunk 1]
```diff
@@ -39,6 +40,30 @@ namespace golos { namespace chain {
...
+        };
+
+        id_type id;
+
+        account_name_type author;
+        shared_string permlink;
+        time_point_sec created;
+        time_point_sec modified;
+        asset specification_cost;
+        time_point_sec specification_deadline;
```
[the remark this answers is not in the document]
Comment (maslenitsa93) [reply libraries/chain/include/golos/chain/worker_proposal_objects.hpp]: @zxcat in client application will be `_deadline` anyway. With `_eta` user should do calculations each time when he wants check deadline, or calculate once and write it to notepad :) Research any freelance website, deadline is anywhere.

File: libraries/chain/steem_evaluator.cpp [hunk 2]
```diff
@@ -456,11 +456,19 @@ namespace golos { namespace chain {
...
-                    logic_exception::cannot_delete_comment_with_worker_proposal,
-                    "Cannot delete a comment with worker proposal.");
+                if (comment.parent_author == STEEMIT_ROOT_POST_PARENT) {
+                    const auto& wpo_idx = db().get_index<worker_proposal_index, by_permlink>();
+                    auto wpo_itr = wpo_idx.find(std::make_tuple(o.author, o.permlink));
+                    GOLOS_CHECK_LOGIC(wpo_itr == wpo_idx.end(),
+                        logic_exception::cannot_delete_comment_with_worker_proposal,
+                        "Cannot delete a comment with worker proposal.");
+                } else {
+                    const auto& wto_idx = _db.get_index<worker_techspec_index, by_permlink>();
```
Comment (afalaleev) [inline libraries/chain/steem_evaluator.cpp]: Attach tspec to a post too.

File: libraries/chain/worker_proposal_evaluators.cpp [hunk 3]
```diff
@@ -58,4 +59,76 @@ namespace golos { namespace chain {
...
+
+        const auto& wto_idx = _db.get_index<worker_techspec_index, by_permlink>();
+        auto wto_itr = wto_idx.find(std::make_tuple(o.author, o.permlink));
+        if (wto_itr == wto_idx.end()) {
+            GOLOS_THROW_MISSING_OBJECT("worker_techspec_object", fc::mutable_variant_object()("author",o.author)("permlink",o.permlink));
+        }
+
+        // TODO: check wpo has not this techspec as approved
+        //const auto& wpo_idx = _db.get_index<worker_proposal_index, by_permlink>();
+        //auto wpo_itr = wpo_idx.find(std::make_tuple(comment.parent_author, comment.parent_permlink));
```
Comment (afalaleev) [inline libraries/chain/worker_proposal_evaluators.cpp]: You shouldn't add limits on delete tspec.

You should unfreeze asset for approved tspec if author want to delete it.

File: libraries/chain/worker_proposal_evaluators.cpp [hunk 4]
```diff
@@ -55,7 +56,89 @@ namespace golos { namespace chain {
             logic_exception::cannot_delete_worker_proposal_with_premade_work,
             "Cannot delete worker proposal with premade work");
 
+        const auto& wto_idx = _db.get_index<worker_techspec_index, by_worker_proposal>();
+        auto wto_itr = wto_idx.find(std::make_tuple(o.author, o.permlink));
+        while (wto_itr != wto_idx.end()) {
+            const auto& cur_wto = *wto_itr;
+            ++wto_itr;
+            _db.remove(cur_wto);
+        }
```
[the remark this answers is not in the document]
Comment (maslenitsa93) [reply libraries/chain/worker_proposal_evaluators.cpp]: @afalaleev https://github.com/GolosChain/golos.worker/blob/master/contracts/golos.worker/golos.worker.cpp#L620-L635

File: libraries/chain/steem_evaluator.cpp [hunk 5]
```diff
@@ -455,12 +455,20 @@ namespace golos { namespace chain {
...
                 auto wpo_itr = wpo_idx.find(std::make_tuple(o.author, o.permlink));
                 GOLOS_CHECK_LOGIC(wpo_itr == wpo_idx.end(),
                     logic_exception::cannot_delete_comment_with_worker_proposal,
                     "Cannot delete a comment with worker proposal.");
+
+                const auto& wto_idx = _db.get_index<worker_techspec_index, by_permlink>();
+                auto wto_itr = wto_idx.find(std::make_tuple(o.author, o.permlink));
+                GOLOS_CHECK_LOGIC(wto_itr == wto_idx.end(),
+                    logic_exception::cannot_delete_comment_with_worker_techspec,
+                    "Cannot delete a comment with worker techspec.");
```
[the remark this answers is not in the document]
Comment (maslenitsa93) [reply libraries/chain/steem_evaluator.cpp]: @afalaleev And should not ban deleting of post which has tspecs? Why?

File: libraries/chain/worker_proposal_evaluators.cpp [hunk 6]
```diff
@@ -47,15 +48,108 @@ namespace golos { namespace chain {
...
+                wto.payments_count = o.payments_count;
+                wto.payments_interval = o.payments_interval;
+            });
+
+            return;
+        }
+
+        _db.create<worker_techspec_object>([&](worker_techspec_object& wto) {
+            wto.author = o.author;
+            wto.permlink = comment.permlink;
```
Comment (afalaleev) [inline libraries/chain/worker_proposal_evaluators.cpp]: `from_string()`

**Pull request merged**